### PR TITLE
Use a custom `decorateStory` function in React

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: #12255 #11657, probably more

## What I did

To fix issue with stories remounting when re-rendering. The issue was the way the `decorateStory` function created a new "bound" story function each time you called it.

This meant React wasn't able to recognize the fact that we were just changing props to the story rather than rendering a whole new component.

The fix is to do the "binding" via React context rather than by creating a new bound function on each call.

Also, tweaked the loaders implementation so React could avoid rebinding the decorated story function with the loaded value each time.

## How to test

I added a story that renders "2" rather than the previous "4" value ;) Not sure if we have the test harness in place to check this via a proper react-rendered Jest test?

## NOTES:

 - I guess this won't work in React 15 (`createContext` didn't exist). What should we do here? I am leaning toward just using the default decorate story and not fixing the bug 🤷 

 - Type changes might be a bit funky, open to improvements.

### CODE REVIEW

File: app/react/src/client/preview/render.tsx [hunk 1]
```diff
@@ -45,19 +52,42 @@ class ErrorBoundary extends Component<{
...
+export const applyDecorator = (story: StoryFn, decorator: DecoratorFunction) => {
+  const Story = story as FunctionComponent<StoryContext>;
+  // You cannot override the parameters key, it is fixed
+  const BoundStory = (
+    { parameters, ...innerStoryContext }: StoryContext | { parameters: {} } = { parameters: {} }
+  ) => {
+    return (
+      <StoryContextContext.Consumer>
+        {(storyContext) => <Story {...{ ...storyContext, ...innerStoryContext }} />}
+      </StoryContextContext.Consumer>
```
Comment (Sawtaytoes) [inline app/react/src/client/preview/render.tsx:66]: I'm curious why we're not using `useContext` instead of render props.

**Comment by Sawtaytoes** [comment]
I just put up an issue on this same thing:
https://github.com/storybookjs/storybook/issues/14475

I think this issue is a bigger deal. It also explains a lot of other issues I'd experienced in the past; things that I assume my code is doing wrong, not Storybook.

Another issue talking about an infinite `useEffect` also defines a custom `decorators` function. I'm assuming that's this person's issue as well:
https://github.com/storybookjs/storybook/issues/12435

**Comment by Sawtaytoes** [comment]
> The fix is to do the "binding" via React context rather than by creating a new bound function on each call.

You don't have to use React context, but I agree, you definitely want to do the binding off cycle.

If you're concerned about React 15, use a class component and legacy context. Both are still available even in React 17, but this also means, at the point the React team decides to remove legacy context or class components, this code will need to be rewritten.

Another option is passing the function or functions down through props. It's gonna work no matter what React version if it's just props.

File: app/react/src/client/preview/render.tsx (hunk 1 above)
Comment (tmeasday) [reply app/react/src/client/preview/render.tsx:66]: I can't quite remember, but my guess would be that `Context.Consumer` is an older API than `useContext`.

**Comment by tmeasday** [comment]
Thanks for your comments @Sawtaytoes 

> Another option is passing the function or functions down through props. It's gonna work no matter what React version if it's just props.

This isn't really an option though because we don't have control over the decorators. We can't guarantee the user will pass the props down through any custom decorators they write.

In the simple case -- suppose the user defines a single decorator:

```js
const decorator = (StoryFn, context) => <StoryFn/>;
```

Then by default the `StoryFn` they are passed here is going to get rendered with no props.

Now we can make `StoryFn` whaterver we like at "story definition time":

```js
const RealStoryFn = /*...*/;
const StoryFn = () => <RealStoryFn context={...} />
```

The problem is that `context` here (the SB context) can change on each invocation of the story. So we don't know what `context` is at this point, we need to make Story function "get" it from somewhere. That's what I've used React context for. I'm open to other ideas (although I think this is fine?)

**Comment by Sawtaytoes** [comment]
I agree. Instead of passing props, context is the better choice. But also like you said, not available in React 15. Is that an issue? I think Storybook is on React 17 which allows multiple versions of React to run in parallel.

## React context

And you're right, `Context.Consumer` is an older protocol. It's the pre-hooks way of using context at the cost of performance. Render-props, in general, incurs a steep performance penalty when used in aggregate because it prevents memoization.

In smaller render jobs, it's less noticeable. When doing a lot of updates (like a slider or a whole app), it's best to prevent re-renders using memoization which is where `useContext` comes in.

I don't know how this affects Storybook's performance.

## The bug

This bug has caused me many problems today and in the past. I always assumed it was something I was doing wrong.

I'd personally like to see it fixed immediately ;).

I often use controls to update component state and do much of my UX development in Storybook. This is actually something I've promoted at work too.

Having Storybook's controls behave like an app would makes it extremely valuable for both manual and automated testing.

**Comment by shilman** [comment]
@Sawtaytoes FYI, we're in a merge freeze while the latest release is stabilizing. We'll resume merging PRs and publishing daily-ish alphas in a week or two. Thanks for your patience!

**Comment by shilman** [comment]
@tmeasday can you look at the failing unit test?

**Comment by shilman** [comment]
@tmeasday is this still on your radar?

**Comment by tmeasday** [comment]
@shilman I have been back and forth on this one a bit.

Firstly I implemented it with a React Context (as discussed above), but the downside is that the story needs to be rendered inside a context `<Provider>` which stuffs up things like shallow rendering or virtual-dom based snapshotting in general.

Secondly I tried an approach using a global var for the story context, but in discussions with @ndelangen we released this will potentially have problems if you render two stories at once in the same `ReactDom.render` call[1], as React can interleave the rendering of two components at once. I guess this is why React Context exists in the first place.

There might be a smarter way to solve this, @ndelangen were going to look at it together.

[1] I'm not sure if the same problem applies to SB hooks btw, I suspect it might.

**Comment by tmeasday** [comment]
Discussed this with @ndelangen. Our proposed solution is to add a new `store.getStoryContext(id)` method, and bind it to the decorators with something like:

```js
decorateStory(storyFn, decorator, getContext) {
  const storyWithContext = (innerContext) => storyFn({getContext(), ...innerContext });

  return context => decorator(storyFn, context);
}
```

We can then refactor the store to do something like:

```js
const decoratedStoryFn = applyDecorators(decorators, originalStoryFn, () => this.getStoryContext(id));

storyFn = () => {
  this._contexts[id] = { ... };

  const result = decoratedStoryFn();

  this._contexts[id] = null;
  return result;
}

getStoryContext(id) {
  return this._contexts[id];
}
```

**Comment by tmeasday** [comment]
Closing in favour of https://github.com/storybookjs/storybook/pull/14677 which implements the above.

**Pull request closed without merging**